include/linux/sysfs.h

```diff
@@ -7,7 +7,7 @@
 #include <linux/mm.h>
 #include_next <linux/sysfs.h>
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 10, 0) && RHEL_RELEASE_CODE < 0x805
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 4, 104) || LINUX_VERSION_CODE >= KERNEL_VERSION(5, 5, 0) && LINUX_VERSION_CODE < KERNEL_VERSION(5, 10, 0)) && RHEL_RELEASE_CODE < 0x805
```

Comment:
If the first condition is true, then the third is implicitly true. It might be clearer to move the open parenthesis to the second condition:

#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 4, 104) || (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 5, 0) && LINUX_VERSION_CODE < KERNEL_VERSION(5, 10, 0)) && RHEL_RELEASE_CODE < 0x805

Also, I have been wondering about the use of RHEL_RELEASE_CODE for Ubuntu builds. I'm assuming these are undefined? Assuming we have everything else correct for an Ubuntu build, would it break if the RHEL_RELEASE_CODE condition was reversed? (i.e. RHEL_RELEASE_CODE >= 0x805)

I may be paranoid - but I'm worried that the use of RHEL_RELEASE_CODE in Ubuntu builds might lead to some issues.

Reply:
You are right it shall say sysfs_emit().

On Ubuntu RHEL_RELEASE_CODE is undefined: and gives warning: "RHEL_RELEASE_CODE" is not defined, evaluates to 0.
So  RHEL_RELEASE_CODE < 0x805 is evaluated to true for Ubuntu. It is also only evaluated if the expression before "&& RHEL..." is true. We are using this construct in a number of places in the backport drivers to have criteria that works for both Ubuntu and RHEL.

We need the parenthesis around the expression to the left of "&& RHEL..." because it needs to evaluate to true for RHEL 8.x that have kernel 4.18.0-xxx in order for the RHEL_RELEASE_CODE < 0x805 to take effect. We could add the parenthesis you suggest for clarity. I have force pushed an update.